File: frontend/src/types/initiatives.ts
/**
 * TypeScript types for the initiatives API.
 * Mirrors backend/src/schemas/initiatives.py and initiatives-api.md.
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type InitiativeStatus =
  | "idea"
  | "planned"
  | "approved"
  | "in_progress"
  | "completed"
  | "rejected";

export type InitiativeType = "custom" | "ai_suggested";

export type ConfidenceLevel = "high" | "medium" | "low";

/** Valid next statuses for each current status (mirrors backend state machine). */
export const STATUS_TRANSITIONS: Record<InitiativeStatus, InitiativeStatus[]> = {
  idea: ["planned", "rejected"],
  planned: ["approved", "rejected"],
  approved: ["in_progress", "rejected"],
  in_progress: ["completed", "rejected"],
  completed: [],
  rejected: [],
};

// ---------------------------------------------------------------------------
// Sub-objects
// ---------------------------------------------------------------------------

export interface EmissionSourceLink {
  id: string;
  activity: string;
  question: string;
  question_group: string;
  answer_unit: string | null;
  co2e_tonnes: number;
  company_unit_name: string | null;
  company_unit_type: string | null;
}

export interface ScenarioRef {
  id: string;
  name: string;
  is_baseline: boolean;
}

export interface InitiativeWarning {
  code: string;
  message: string;
}

// ---------------------------------------------------------------------------
// Create / Update payloads
// ---------------------------------------------------------------------------

export interface InitiativeCreate {
  name: string;
  description?: string | null;
  initiative_type?: InitiativeType;
  status?: InitiativeStatus;
  capex_gbp: number;
  opex_annual_gbp?: number | null;
  co2e_reduction_annual_tonnes: number;
  lifespan_years?: number;
  owner?: string | null;
  confidence?: ConfidenceLevel | null;
  notes?: string | null;
  emission_source_ids: string[];
}

export interface InitiativeUpdate {
  name?: string;
  description?: string | null;
  initiative_type?: InitiativeType;
  status?: InitiativeStatus;
  capex_gbp?: number;
  opex_annual_gbp?: number | null;
  co2e_reduction_annual_tonnes?: number;
  lifespan_years?: number | null;
  owner?: string | null;
  confidence?: ConfidenceLevel | null;
  notes?: string | null;
  emission_source_ids?: string[];
}

export interface StatusUpdate {
  status: InitiativeStatus;
  notes?: string | null;
}

// ---------------------------------------------------------------------------
// Responses
// ---------------------------------------------------------------------------

/** Full initiative detail (POST 201, GET detail, PUT, PATCH /status). */
export interface Initiative {
  id: string;
  name: string;
  description: string | null;
  initiative_type: InitiativeType;
  status: InitiativeStatus;
  capex_gbp: number;
  opex_annual_gbp: number | null;
  co2e_reduction_annual_tonnes: number;
  cost_per_tonne: number;
  payback_years: number | null;
  lifespan_years: number;
  owner: string | null;
  confidence: ConfidenceLevel | null;
  notes: string | null;
  source_suggestion_id: string | null;
  emission_sources: EmissionSourceLink[];
  scenarios: ScenarioRef[];
  warnings: InitiativeWarning[];
  created_at: string;
  updated_at: string;
}

/** Summary row in the paginated list. */
export interface InitiativeListItem {
  id: string;
  name: string;
  initiative_type: InitiativeType;
  status: InitiativeStatus;
  capex_gbp: number;
  co2e_reduction_annual_tonnes: number;
  cost_per_tonne: number;
  owner: string | null;
  confidence: ConfidenceLevel | null;
  emission_source_count: number;
  scenario_count: number;
  created_at: string;
}

export interface InitiativeListResponse {
  items: InitiativeListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// ---------------------------------------------------------------------------
// Overlap
// ---------------------------------------------------------------------------

export interface SharedSourceInfo {
  id: string;
  activity: string;
  question_group: string;
  co2e_tonnes: number;
}

export interface OverlappingInitiativeInfo {
  id: string;
  name: string;
  shared_sources: SharedSourceInfo[];
  combined_reduction_pct: number;
  warning: string;
}

export interface OverlapResponse {
  initiative_id: string;
  overlapping_initiatives: OverlappingInitiativeInfo[];
  total_overlap_co2e_tonnes: number;
}

// ---------------------------------------------------------------------------
// Bulk validate
// ---------------------------------------------------------------------------

export interface ValidationIssue {
  initiative_id: string;
  issue_type: string;
  message: string;
  source_id?: string | null;
  source_name?: string | null;
}

export interface BulkValidateResponse {
  valid: boolean;
  issues: ValidationIssue[];
}

// ---------------------------------------------------------------------------
// MACC chart data
// ---------------------------------------------------------------------------

export interface MACCBar {
  initiative_id: string;
  name: string;
  status: InitiativeStatus;
  initiative_type: InitiativeType;
  confidence: ConfidenceLevel | null;
  owner: string | null;
  capex_gbp: number;
  co2e_reduction_annual_tonnes: number;
  cost_per_tonne: number;
  opex_annual_gbp: number | null;
  /** Cumulative start x-position (abatement axis). */
  x_start: number;
  /** Cumulative end x-position. */
  x_end: number;
  bar_width: number;
  bar_height: number;
  is_negative_cost: boolean;
}

export interface MACCSummary {
  total_capex_gbp: number;
  total_co2e_reduction_annual_tonnes: number;
  weighted_avg_cost_per_tonne: number;
  bar_count: number;
  negative_cost_count: number;
  positive_cost_count: number;
  max_abatement_potential: number;
}

export interface MACCData {
  bars: MACCBar[];
  summary: MACCSummary;
}

// ---------------------------------------------------------------------------
// Filters / sort
// ---------------------------------------------------------------------------

export interface InitiativeFilters {
  status?: string;            // comma-separated
  initiative_type?: InitiativeType;
  scope?: number;
  sort_by?: "cost_per_tonne" | "co2e_reduction" | "cost" | "name" | "created_at";
  sort_order?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

// ---------------------------------------------------------------------------
// Cascade source-selection types
// ---------------------------------------------------------------------------

export interface CascadeScopeItem {
  scope: number;
  co2e_tonnes: number;
  source_count: number;
}

export interface CascadeScopeResponse {
  items: CascadeScopeItem[];
}

export interface CascadeQuestionGroupItem {
  question_group: string;
  co2e_tonnes: number;
  source_count: number;
}

export interface CascadeQuestionGroupResponse {
  scope: number;
  items: CascadeQuestionGroupItem[];
}

export interface CascadeQuestionItem {
  question: string;
  co2e_tonnes: number;
  source_count: number;
}

export interface CascadeQuestionResponse {
  scope: number;
  question_group: string;
  items: CascadeQuestionItem[];
}

export interface CascadeActivityItem {
  activity: string;
  co2e_tonnes: number;
  source_count: number;
}

export interface CascadeActivityResponse {
  scope: number;
  question_group: string;
  question: string;
  items: CascadeActivityItem[];
}

export interface CascadeCompanyUnitItem {
  emission_source_id: string;
  company_unit_id: string;
  company_unit_name: string;
  company_unit_type: string | null;
  co2e_tonnes: number;
}

export interface CascadeCompanyUnitResponse {
  activity: string;
  items: CascadeCompanyUnitItem[];
}

File: backend/src/schemas/initiatives.py
"""Pydantic schemas for initiative endpoints.

Covers: InitiativeCreate, InitiativeUpdate, InitiativeResponse,
InitiativeListResponse, InitiativeListItem, StatusUpdate,
OverlapResponse, BulkValidateResponse, CascadeResponse variants
— per initiatives-api.md and emissions-api.md.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


# ---------------------------------------------------------------------------
# Enums / literal values (kept as plain strings for Pydantic v2 compatibility)
# ---------------------------------------------------------------------------

VALID_STATUSES = {"idea", "planned", "approved", "in_progress", "completed", "rejected"}
STATUS_TRANSITIONS: dict[str, set[str]] = {
    "idea": {"planned", "rejected"},
    "planned": {"approved", "rejected"},
    "approved": {"in_progress", "rejected"},
    "in_progress": {"completed", "rejected"},
    "completed": set(),
    "rejected": set(),
}
VALID_CONFIDENCE = {"high", "medium", "low"}
VALID_TYPES = {"custom", "ai_suggested"}


# ---------------------------------------------------------------------------
# Linked source sub-objects
# ---------------------------------------------------------------------------


class EmissionSourceLink(BaseModel):
    """Minimal source info returned inside an initiative response."""

    id: str
    activity: str
    question: str
    question_group: str
    answer_unit: str | None = None
    co2e_tonnes: float
    company_unit_name: str | None = None
    company_unit_type: str | None = None


class ScenarioRef(BaseModel):
    """Minimal scenario reference returned inside an initiative response."""

    id: str
    name: str
    is_baseline: bool


# ---------------------------------------------------------------------------
# Warning sub-object
# ---------------------------------------------------------------------------


class InitiativeWarning(BaseModel):
    """Non-blocking warning attached to an initiative response."""

    code: str
    message: str


# ---------------------------------------------------------------------------
# Create / Update
# ---------------------------------------------------------------------------


class InitiativeCreate(BaseModel):
    """Body for POST /initiatives."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    initiative_type: str = Field("custom")
    status: str = Field("idea")
    capex_gbp: float = Field(..., ge=0)
    opex_annual_gbp: float | None = None
    co2e_reduction_annual_tonnes: float = Field(..., gt=0)
    lifespan_years: int = Field(10, ge=1)
    owner: str | None = Field(None, max_length=255)
    confidence: str | None = None
    notes: str | None = None
    emission_source_ids: list[str] = Field(..., min_length=1)

    @field_validator("initiative_type")
    @classmethod
    def _validate_type(cls, v: str) -> str:
        if v not in VALID_TYPES:
            raise ValueError(f"initiative_type must be one of {VALID_TYPES}")
        return v

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: str) -> str:
        if v not in VALID_STATUSES:
            raise ValueError(f"status must be one of {VALID_STATUSES}")
        return v

    @field_validator("confidence")
    @classmethod
    def _validate_confidence(cls, v: str | None) -> str | None:
        if v is not None and v not in VALID_CONFIDENCE:
            raise ValueError(f"confidence must be one of {VALID_CONFIDENCE}")
        return v


class InitiativeUpdate(BaseModel):
    """Body for PUT /initiatives/{id}.

    All fields optional — partial updates allowed.
    Status transitions are validated by the service layer.
    """

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    initiative_type: str | None = None
    status: str | None = None
    capex_gbp: float | None = Field(None, ge=0)
    opex_annual_gbp: float | None = None
    co2e_reduction_annual_tonnes: float | None = Field(None, gt=0)
    lifespan_years: int | None = Field(None, ge=1)
    owner: str | None = None
    confidence: str | None = None
    notes: str | None = None
    emission_source_ids: list[str] | None = None

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: str | None) -> str | None:
        if v is not None and v not in VALID_STATUSES:
            raise ValueError(f"status must be one of {VALID_STATUSES}")
        return v

    @field_validator("confidence")
    @classmethod
    def _validate_confidence(cls, v: str | None) -> str | None:
        if v is not None and v not in VALID_CONFIDENCE:
            raise ValueError(f"confidence must be one of {VALID_CONFIDENCE}")
        return v


class StatusUpdate(BaseModel):
    """Body for PATCH /initiatives/{id}/status."""

    status: str
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: str) -> str:
        if v not in VALID_STATUSES:
            raise ValueError(f"status must be one of {VALID_STATUSES}")
        return v


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class InitiativeResponse(BaseModel):
    """Full initiative detail — returned by POST, GET /{id}, PUT, PATCH /status."""

    id: str
    name: str
    description: str | None
    initiative_type: str
    status: str
    capex_gbp: float
    opex_annual_gbp: float | None
    co2e_reduction_annual_tonnes: float
    cost_per_tonne: float
    payback_years: float | None
    lifespan_years: int
    owner: str | None
    confidence: str | None
    notes: str | None
    source_suggestion_id: str | None
    emission_sources: list[EmissionSourceLink]
    scenarios: list[ScenarioRef]
    warnings: list[InitiativeWarning] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InitiativeListItem(BaseModel):
    """Summary row — returned inside GET / list."""

    id: str
    name: str
    initiative_type: str
    status: str
    capex_gbp: float
    co2e_reduction_annual_tonnes: float
    cost_per_tonne: float
    owner: str | None
    confidence: str | None
    emission_source_count: int
    scenario_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class InitiativeListResponse(BaseModel):
    """Paginated initiative list."""

    items: list[InitiativeListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------------------------------------------------------------------
# Overlap detection
# ---------------------------------------------------------------------------


class SharedSourceInfo(BaseModel):
    id: str
    activity: str
    question_group: str
    co2e_tonnes: float


class OverlappingInitiativeInfo(BaseModel):
    id: str
    name: str
    shared_sources: list[SharedSourceInfo]
    combined_reduction_pct: float
    warning: str


class OverlapResponse(BaseModel):
    initiative_id: str
    overlapping_initiatives: list[OverlappingInitiativeInfo]
    total_overlap_co2e_tonnes: float


# ---------------------------------------------------------------------------
# Bulk validate
# ---------------------------------------------------------------------------


class BulkValidateRequest(BaseModel):
    initiative_ids: list[str] = Field(..., min_length=1)


class ValidationIssue(BaseModel):
    initiative_id: str
    issue_type: str
    message: str
    source_id: str | None = None
    source_name: str | None = None


class BulkValidateResponse(BaseModel):
    valid: bool
    issues: list[ValidationIssue]


# ---------------------------------------------------------------------------
# Cascade source-selection schemas (emissions-api.md §Cascade)
# ---------------------------------------------------------------------------


class CascadeScopeItem(BaseModel):
    """One scope entry returned by GET /emissions/cascade/scopes."""

    scope: int
    co2e_tonnes: float
    source_count: int


class CascadeScopeResponse(BaseModel):
    items: list[CascadeScopeItem]


class CascadeQuestionGroupItem(BaseModel):
    question_group: str
    co2e_tonnes: float
    source_count: int


class CascadeQuestionGroupResponse(BaseModel):
    scope: int
    items: list[CascadeQuestionGroupItem]


class CascadeQuestionItem(BaseModel):
    question: str
    co2e_tonnes: float
    source_count: int


class CascadeQuestionResponse(BaseModel):
    scope: int
    question_group: str
    items: list[CascadeQuestionItem]


class CascadeActivityItem(BaseModel):
    activity: str
    co2e_tonnes: float
    source_count: int


class CascadeActivityResponse(BaseModel):
    scope: int
    question_group: str
    question: str
    items: list[CascadeActivityItem]


class CascadeCompanyUnitItem(BaseModel):
    """One company-unit leaf entry — carries an emission_source_id for linking."""

    emission_source_id: str
    company_unit_id: str
    company_unit_name: str
    company_unit_type: str | None
    co2e_tonnes: float


class CascadeCompanyUnitResponse(BaseModel):
    activity: str
    items: list[CascadeCompanyUnitItem]